File: 0046-permutations/0046-permutations.cpp
class Solution {
private:
    void findPermute(vector<int> &nums, int start, vector<int> &current, vector<vector<int>> &result) {
        if (current.size() == nums.size()) {
            result.push_back(current);
            return;
        }

        for(int i = start; i < nums.size(); ++i) {
            swap(nums[i], nums[start]);
            current.push_back(nums[start]);

            findPermute(nums, start + 1, current, result);

            current.pop_back();
            swap(nums[i], nums[start]);
        }
    }

public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> current;

        findPermute(nums, 0, current, result);

        return result;
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # backtraking
        # cur: the list to record numbers we choosed
        # start: define the start idx

        # base case: when length of cur == length of nums add the cur to result and return

        # else: start from idx and swap the number we choose


        def backtrack(cur, start, result):
            if start == len(nums):
                result.append(cur[:])
                return

            for i in range(start, len(nums)):
                nums[start], nums[i] = nums[i], nums[start]# swap to make permutation
                cur.append(nums[start])

                backtrack(cur, start + 1, result)

                cur.pop()
                nums[start], nums[i] = nums[i], nums[start]

        result = []
        backtrack([], 0, result)
        
        return result
        
    
